src/theory/quantifiers/instantiate.cpp

```diff
@@ -0,0 +1,831 @@
+                                              bool modEq,
+                                              bool addedLem)
+{
+  if (!addedLem)
+  {
+    // record the instantiation for deletion later
+    d_recorded_inst.push_back(std::pair<Node, std::vector<Node> >(q, terms));
+  }
+  if (options::incrementalSolving())
+  {
+    Trace("inst-add-debug")
+        << "Adding into context-dependent inst trie, modEq = " << modEq
+        << std::endl;
+    inst::CDInstMatchTrie* imt;
+    std::map<Node, inst::CDInstMatchTrie*>::iterator it =
+        d_c_inst_match_trie.find(q);
+    if (it != d_c_inst_match_trie.end())
+    {
+      imt = it->second;
+    }
+    else
+    {
+      imt = new inst::CDInstMatchTrie(d_qe->getUserContext());
+      d_c_inst_match_trie[q] = imt;
+    }
+    d_c_inst_match_trie_dom.insert(q);
+    return imt->addInstMatch(d_qe, q, terms, d_qe->getUserContext(), modEq);
+  }
+  else
+  {
+    Trace("inst-add-debug") << "Adding into inst trie" << std::endl;
+    return d_inst_match_trie[q].addInstMatch(d_qe, q, terms, modEq);
+  }
+}
+
+bool Instantiate::removeInstantiationInternal(Node q, std::vector<Node>& terms)
+{
+  if (options::incrementalSolving())
+  {
+    std::map<Node, inst::CDInstMatchTrie*>::iterator it =
+        d_c_inst_match_trie.find(q);
+    if (it != d_c_inst_match_trie.end())
+    {
+      return it->second->removeInstMatch(q, terms);
+    }
+    return false;
+  }
+  return d_inst_match_trie[q].removeInstMatch(q, terms);
+}
+
+Node Instantiate::getTermForType(TypeNode tn)
+{
+  if (d_qe->getTermEnumeration()->isClosedEnumerableType(tn))
+  {
+    return d_qe->getTermEnumeration()->getEnumerateTerm(tn, 0);
+  }
+  else
+  {
+    return d_qe->getTermDatabase()->getOrMakeTypeGroundTerm(tn);
+  }
+}
+
+bool Instantiate::printInstantiations(std::ostream& out)
+{
+  bool useUnsatCore = false;
+  std::vector<Node> active_lemmas;
+  if (options::trackInstLemmas() && getUnsatCoreLemmas(active_lemmas))
+  {
+    useUnsatCore = true;
+  }
+  bool printed = false;
+  if (options::incrementalSolving())
+  {
+    for (std::pair<const Node, inst::CDInstMatchTrie*>& t : d_c_inst_match_trie)
+    {
+      bool firstTime = true;
+      t.second->print(out, t.first, firstTime, useUnsatCore, active_lemmas);
+      if (!firstTime)
+      {
+        out << ")" << std::endl;
+      }
+      printed = printed || !firstTime;
+    }
+  }
+  else
+  {
+    for (std::pair<const Node, inst::InstMatchTrie>& t : d_inst_match_trie)
+    {
+      bool firstTime = true;
+      t.second.print(out, t.first, firstTime, useUnsatCore, active_lemmas);
+      if (!firstTime)
+      {
+        out << ")" << std::endl;
+      }
+      printed = printed || !firstTime;
+    }
+  }
+  return printed;
+}
+
+void Instantiate::getInstantiatedQuantifiedFormulas(std::vector<Node>& qs)
+{
+  if (options::incrementalSolving())
+  {
+    for (context::CDHashSet<Node, NodeHashFunction>::const_iterator it =
```

Comment:
for-each

Reply:
missed

Reply:
I'm not sure that for-each works with context-dependent data structures (I'd also have to performance test to be sure). That's why I left this one.